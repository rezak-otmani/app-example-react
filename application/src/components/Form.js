import React from 'react'
import { Field, reduxForm } from 'redux-form'

const required = value => value ? undefined : 'Required'
const maxLength = max => value =>
  value && value.length > max ? `Must be ${max} characters or less` : undefined
const maxLength15 = maxLength(15)
const number = value => value && isNaN(Number(value)) ? 'Must be a number' : undefined
const minValue = min => value =>
  value && value < min ? `Must be at least ${min}` : undefined
const minValue18 = minValue(18)
const email = value =>
  value && !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(value) ?
  'Invalid email address' : undefined
const tooOld = value =>
  value && value > 65 ? 'You might be too old for this' : undefined
const aol = value =>
  value && /.+@aol\.com/.test(value) ?
  'Really? You still use AOL for your email?' : undefined

const renderField = ({ input, label, type, meta: { touched, error, warning } }) => (
  <div>
    <label>{label}</label>
    <div>
      <input {...input} placeholder={label} type={type}/>
      {touched && ((error && <span>{error}</span>) || (warning && <span>{warning}</span>))}
    </div>
  </div>
)

const Form = (props) => {
  const { handleSubmit, pristine, reset, submitting } = props
  return (
    <form class="ui form" onSubmit={handleSubmit}>
         <div> 
          <Field 
                   name="name"
                   type="text"
                   placeholder="Nom"
                   label="Nom"
                   component={renderField} 
                   validate={[ required, maxLength15 ]}
           />
        
           <Field
                   name="email" type="email"
                   component={renderField}
                   label="Email"
                   validate={email}
                   warn={aol}
           />

           <Field
                    name="telephone"
                    type="text"
                    component={renderField}
                    label="Téléphone"
                    validate={[ required, maxLength15 ]}
 
           />
                       <Field 
                    name="adresse" 
                    type="text"
                    component={renderField} 
                    label="Adresse"
                    validate={[ required, maxLength15 ]}

           />


      </div>
        <br />
      <div>
        <button class="positive negative massive ui button" type="button" disabled={pristine || submitting} onClick={reset}>Clear Values</button>
        <button class="positive massive ui button" type="submit" disabled={submitting}>Send<i class="right arrow icon"></i></button>
      </div>
    </form>
  )
}

export default reduxForm({
  form: 'Form' // a unique identifier for this form
})(Form)
