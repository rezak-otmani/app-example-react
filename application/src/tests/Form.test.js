import React from 'react';
import Enzyme, { shallow, render, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Form from '../components/Form';



describe('Form Component', () => {
        
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Form />
    );
  });

   


     describe('Render Form Component', () => {


        it('Should render Form without throwing an error', () => {
  expect(shallow(<Form />).exists(<form className='ui form'></form>)).toBe(true)})
               
         it('render the name input', () => {
   	 expect(shallow(<Form />).find('[name="name"]').length), 1
  	 })

         it('render the email input', () => {
  	  expect(shallow(<Form />).find('[name="email"]').length), 1
  	 })

         it('render the subject input', () => {
          expect(shallow(<Form />).find('[name="subject"]').length), 1
         })

         it('render the phone input', () => {
          expect(shallow(<Form />).find('[name="phone"]').length), 1
         })

         it('render the  adresse input', () => {
          expect(shallow(<Form />).find('[name="adresse"]').length), 1
         })

          it('render the <SubmitButton /> component', () => {
          (wrapper.find('button[type="submit"]').length, 1);
          });

           it('renders the  <ResetButton /> component', () => {
          (wrapper.find('button[type="reset"]').length, 1);
          });
     
 });              
        


           describe('The user populates the inputs', () => {
                   const name = 'john';
                   const email = 'example@gmail.com';
                   const subject = 'demo';
                   const phone = '1234567890';
                   const adresse = 'my adresse';


    it('should respond to change event and change the state of the Form Component', () => {
    beforeEach(() => {
      const input1 = wrapper.find('Field[name="name"]').first();
      const input2 = wrapper.find('Field[name="email"]').first();
      const input3 = wrapper.find('Field[name="subject"]').first();
      const input4 = wrapper.find('Field[name="phone"]').first();
      const input5 = wrapper.find('Field[name="adresse"]').first();
      input1.simulate('change', {target: { value: name } });
      input2.simulate('change', {target: { value: email } });
      input3.simulate('change', {target: { value: subject } });
      input4.simulate('change', {target: { value: phone } });
      input5.simulate('change', {target: { value: adresse } });
         expect(wrapper.state().name).toEqual(name);
         expect(wrapper.state().email).toEqual(email);
         expect(wrapper.state().subject).toEqual(subject);
         expect(wrapper.state().phone).toEqual(phone);
         expect(wrapper.state().adresse).toEqual(adresse);

    });
});      
    
    
     describe('and then submits the form', () => {
      beforeEach(() => {
        const form = wrapper.find('form').first();
        form.simulate('submit', {
          preventDefault: () => {},
        });
      });      
        
         });



             




});




});
