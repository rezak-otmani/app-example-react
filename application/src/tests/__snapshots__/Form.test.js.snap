// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="center aligned column"
>
  <form
    className="ui form"
    onSubmit={[Function]}
  >
    <div>
      <input
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        placeholder="Nom"
        type="text"
        value=""
      />
    </div>
    <br />
    <div>
      <input
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        placeholder="Email"
        type="email"
        value=""
      />
    </div>
    <br />
    <div>
      <input
        name="subject"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        placeholder="Sujet"
        type="text"
        value=""
      />
    </div>
    <br />
    <div>
      <input
        name="phone"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        placeholder="Téléphone"
        type="text"
        value=""
      />
    </div>
    <br />
    <div>
      <input
        name="adresse"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        placeholder="Adresse"
        type="text"
        value=""
      />
    </div>
    <br />
    <div>
      <button
        className="massive ui button"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Clear Values
      </button>
      <button
        className="positive massive ui button"
        disabled={false}
        type="submit"
      >
        Send
        <i
          className="right arrow icon"
        />
      </button>
    </div>
  </form>
</div>
`;
